ring/ingester_lifecycle.go

```diff
@@ -16,125 +16,175 @@
 package ring
 
 import (
-	"encoding/json"
 	"fmt"
-	"hash/fnv"
 	"math/rand"
 	"net"
 	"os"
+	"sort"
 	"sync"
 	"time"
 
 	"github.com/prometheus/common/log"
 )
 
 const (
-	infName = "eth0"
+	infName           = "eth0"
+	consulKey         = "ring"
+	heartbeatInterval = 5 * time.Second
 )
 
 // IngesterRegistration manages the connection between the ingester and Consul.
 type IngesterRegistration struct {
-	consul ConsulClient
-	id     string
-	desc   []byte
+	consul    ConsulClient
+	numTokens int
 
-	quit chan struct{}
-	wait sync.WaitGroup
+	id       string
+	hostname string
+	quit     chan struct{}
+	wait     sync.WaitGroup
 }
 
 // RegisterIngester registers an ingester with Consul.
 func RegisterIngester(consulClient ConsulClient, listenPort, numTokens int) (*IngesterRegistration, error) {
-	desc, err := describeLocalIngester(listenPort, numTokens)
+	hostname, err := os.Hostname()
 	if err != nil {
 		return nil, err
 	}
-	buf, err := json.Marshal(desc)
+
+	addr, err := getFirstAddressOf(infName)
 	if err != nil {
 		return nil, err
 	}
 
 	r := &IngesterRegistration{
-		consul: consulClient,
-		id:     desc.ID,
-		desc:   buf,
-		quit:   make(chan struct{}),
+		consul:    consulClient,
+		numTokens: numTokens,
+
+		id: hostname,
+		// hostname is the ip+port of this instance, written to consul sp
+		// the distributors know where to connect.
+		hostname: fmt.Sprintf("%s:%d", addr, listenPort),
+		quit:     make(chan struct{}),
 	}
+
 	r.wait.Add(1)
-	go r.updateLoop()
+	go r.loop()
 	return r, nil
 }
 
-func (r *IngesterRegistration) updateLoop() {
+// Unregister removes ingester config from Consul
+func (r *IngesterRegistration) Unregister() {
+	log.Info("Removing ingester from consul")
+	close(r.quit)
```

Comment:
> Closing the quit channel triggers the unregistration?

Yes

> If so, maybe add a comment, since it relies on action at a distance.

Its just an implementation detail, and keeps the knowledge of tokens (and state transistion) all in one place.  I thought it was quite elegant!

> This call will block until it is unregistered? Either way, please document that in the doc comment.

Yes & will do.